### DISCUSSION TITLE
pixi add option that takes in input a Python module name instead of conda package name

### DISCUSSION THREAD

**Discussion by traversaro** [prefix-dev/pixi#958] [Ideas]
### Problem description

A common workflow I am experiencing working with pixi is:
* Setup a pixi.toml
* Run stuff
* Get an error like "ModuleNotFoundError: No module named 'glfw'"
* Call `pixi add <condapkgname>`, where `<condapkgname>` is the conda-forge package that provides the module not found (in case of glfw, it is pyglfw). 

It would be cool (but I do not know how feasible) to have a `pixi add-pymod` or `pixi add --pymod` command that:
* Query some kind of database that maps python modules names (not python package names) to conda-forge package names (I think there is something similar at the conda-forge level), and then `pixi add` the corresponding conda-forge package name.

Probably it is not a feature that can be implemented right away, but I wanted to track this idea somewhere.

Even more futuristic, if [cps](https://github.com/cps-org/cps) gets actually adopted by the C++ community, it would be cool to do something similar in C++ with cps package names.

**Comment by ruben-arts** [comment]
Interesting idea, we are working on improving conda-python mapping thus this data might be available soon. Not sure if this would fit into `pixi` directly but would be cool to do a prototype!

**Comment by traversaro** [comment]
> Even more futuristic, if [cps](https://github.com/cps-org/cps) gets actually adopted by the C++ community, it would be cool to do something similar in C++ with cps package names.

Probably a bit OT, but just to note it somewhere: without going too much in the future with standards that are not even out in alpha, something else quite useful could be a CMake package --> conda package resolver (fyi @AleTarsi).

**Comment by ruben-arts** [comment]
Ping @wolfv @baszalmstra @nichmor. Do you think this would be easy to add to the `pypi-conda` mapping tool?

**Comment by traversaro** [comment]
> Probably a bit OT, but just to note it somewhere: without going too much in the future with standards that are not even out in alpha, something else quite useful could be a CMake package --> conda package resolver (fyi @AleTarsi).

Again, a bit OT, but just to past in public a comment in a private chat:

> A first pass at the CMake mapping is relatively easy.  If a conda package provides a <PackageName>-config.cmake file, then it provides the PackageName. If a conda package provides a <packagename>-config.cmake file, then it provides all possible packages obtained by varying the cases of packagename, i.e. PackageName, PACKAGENAME, PaCkAgEnAmE, etc etc. For the packages provided by Find<PackageName>.cmake scripts in CMake itself, the simplest solution is just hardcoding. For the packages provided by third-party Find<PackageName>.cmake, the solution is hardcoding ot just ignore them.

**Comment by ruben-arts** [comment]
The CPS feature would be an awesome addition. If that becomes a good standard we would be open to add this as a feature. The python mapping **does not** include the python module name (miss understanding on my part) so this might require an extra mapping step we're currently not working on. Is there already a opensource mapping for this that anyone knows of?

**Comment by traversaro** [comment]
> Is there already a opensource mapping for this that anyone knows of?

I am not an expert of that part of conda-forge, but unless I am missing something it seems that the existing mapping used by grayskull contains both the pypi package name and python module import name, see:
* https://github.com/regro/cf-graph-countyfair/blob/master/mappings/pypi/name_mapping.yaml
* https://github.com/regro/cf-graph-countyfair/blob/master/mappings/pypi/grayskull_pypi_mapping.yaml

**Comment by traversaro** [comment]
marimo has been working on such functionality (see https://twitter.com/marimo_io/status/1774168957739827442), but at the moment I guess they are not handling the case in which the python module is not called as the conda package https://github.com/marimo-team/marimo/issues/1071 .
